lib/protoparser/zabbixconnector/parser.go

```diff
@@ -0,0 +1,315 @@
+	}
+	tag = r.addTag()
+	tag.Key = append(tag.Key[:0], []byte("hostname")...)
+	tag.Value = append(tag.Value[:0], v...)
+
+	v = o.GetStringBytes("name")
+	if len(v) == 0 {
+		return fmt.Errorf("missing `name` element")
+	}
+	tag = r.addTag()
+	tag.Key = append(tag.Key[:0], []byte("__name__")...)
+	tag.Value = append(tag.Value[:0], v...)
+
+	n, err := getFloat64(o, "value")
+	if err != nil {
+		return fmt.Errorf("missing `value` element, %s", err)
+	}
+	r.Value = n
+
+	cl, err := getInt64(o, "clock")
+	if err != nil {
+		return fmt.Errorf("missing `clock` element, %s", err)
+	}
+	ns, err := getInt64(o, "ns")
+	if err != nil {
+		return fmt.Errorf("missing `ns` element, %s", err)
+	}
+	// clock - Number of seconds since Epoch to the moment when value was collected (integer part).
+	// ns - Number of nanoseconds to be added to clock to get a precise value collection time.
+	//
+	// See https://www.zabbix.com/documentation/current/en/manual/appendix/protocols/real_time_export#item-values
+	r.Timestamp = cl*1e3 + ns/1e6
+
+	if *addGroups {
+		groups, err := getArray(o, "groups")
+		if err != nil {
+			return fmt.Errorf("missing `groups` element, %s", err)
+		}
+		for _, g := range groups {
+			k := g.GetStringBytes()
+			if len(k) == 0 {
+				continue
+			}
+
+			tag = r.addTag()
+			tag.Key = append(tag.Key[:0], []byte("group_")...)
+			tag.Key = append(tag.Key, k...)
+			tag.Value = append(tag.Value[:0], "true"...)
+		}
+	}
+
+	tagValuesPlaceholder := []byte(*tagValuePlaceholder)
+	tagsSeparator := []byte(*duplicateTagsSeparator)
+
+	itemTags, err := getArray(o, "item_tags")
+	if err != nil {
+		return fmt.Errorf("missing `item_tags` element, %s", err)
+	}
+
+	if len(tagsSeparator) == 0 { // Do not merge tags
+		for _, t := range itemTags {
+			k := t.GetStringBytes("tag")
+			if len(k) == 0 {
+				continue
+			}
+
+			v := t.GetStringBytes("value")
+			if len(v) == 0 && len(tagValuesPlaceholder) == 0 {
+				continue
+			}
+			tag = r.addTag()
+			tag.Key = append(tag.Key[:0], []byte("tag_")...)
+			tag.Key = append(tag.Key, k...)
+			if len(v) == 0 {
+				tag.Value = append(tag.Value[:0], tagValuesPlaceholder...)
+			} else {
+				tag.Value = append(tag.Value[:0], v...)
+			}
+		}
+	} else { // Merge Tags
+		mapTags := make(map[string][]byte)
+		for _, t := range itemTags {
+			k := t.GetStringBytes("tag")
+			if len(k) == 0 {
+				continue
+			}
+
+			v := t.GetStringBytes("value")
+			if len(v) == 0 && len(tagValuesPlaceholder) == 0 {
+				continue
+			}
+			sk := bytesutil.ToUnsafeString(k)
+			if _, ok := mapTags[sk]; !ok {
+				if len(v) == 0 {
+					mapTags[sk] = tagValuesPlaceholder
+				} else {
+					mapTags[sk] = v
+				}
+			} else {
+				mapTags[sk] = append(mapTags[sk], tagsSeparator...)
+				if len(v) == 0 {
+					mapTags[sk] = append(mapTags[sk], tagValuesPlaceholder...)
+				} else {
+					mapTags[sk] = append(mapTags[sk], v...)
+				}
+			}
+		}
+
+		for k, v := range mapTags {
```

Comment:
golang map produces randomly ordered results. It may lead to inefficient cache usage usage at vmstorage side. So, it's better to sort it tags after map append.